File: huffman0.rb
s = 'Hello world'

def build_tree(s)
    a = s.chars.uniq.map {|c| {char: c, count: s.chars.count(c)}}.sort {|a, b| [b[:count], a[:char]] <=> [a[:count], b[:char]]}
    pq = a.map { |c| {char: c[:char], count: c[:count], left: nil, right: nil} }

    while pq.size > 1 do
        left, right = pq.pop(2)

        if left[:count] < right[:count]
            left, right = right, left
        end

        pq << {char: nil, count: left[:count] + right[:count], left: left, right: right}

        pq.sort! {|a,b| [b[:count], a[:char] || ''] <=> [a[:count], b[:char] || '']}
    end

    pq[0]
end

def build_table(tree)
    q = [{node: tree, code: ''}]

    codes = {}

    while q.size > 0 do
        e = q.shift

        if e[:node][:left].nil? && e[:node][:right].nil?
            codes[e[:node][:char]] = e[:code]
            next
        end

        q.push({node: e[:node][:left], code: e[:code] + '0'}) if !e[:node][:left].nil?

        q.push({node: e[:node][:right], code: e[:code] + '1'}) if !e[:node][:right].nil?
    end

    codes
end

def encode(s)
    tree = build_tree(s)
    codes = build_table(tree)
    s.chars.map {|c| codes[c]}.join.chars.each_slice(8).map {|slice| slice.join.to_i(2).chr}.join
end

def decode(s, tree)
    s1 = s.chars.map {|c| c.ord.to_s(2).rjust(8, '0')}.join
    puts s1
end

s = 'Hello world'
tree = build_tree(s)
codes = build_table(tree)
encoded = encode(s)

puts "Codes: #{codes.inspect}"
puts "Encoded (chars):  #{s.chars.map {|c| c.ljust(codes[c].length, ' ')}.join(' ')}"
puts "Encoded (each):   #{s.chars.map {|c| codes[c]}.join(' ')}"
puts "Encoded (bytes):  " + s.chars.map {|c| codes[c]}.join.chars.each_slice(8).map {|slice| slice.join}.join(' ')
puts "Encoded (hex):    " + s.chars.map {|c| codes[c]}.join.chars.each_slice(8).map {|slice| "0x#{slice.join.to_i(2).to_s(16).rjust(2, '0').upcase}".ljust(8, ' ')}.join(' ')
puts "Encoded (bin):    " + s.chars.map {|c| codes[c]}.join
puts "Encoded (raw):    #{encoded}"

decoded = decode(encoded, tree)
puts "Decoded: #{decoded}"

# bin 10100110101111000001110000010010
# enc 10100110101111000001110000010010

File: huffman0.cpp
#include <string>
#include <map>
#include <queue>
#include <iostream>

struct Node {
    char c;
    int count;
    Node* left;
    Node* right;
};

struct CodeNode {
    Node* node;
    std::string code;
};

int main() {
    const std::string s = "Hello world";

    std::map<char, int> freq;

    for (auto i : s) {
        if (freq.find(i) == freq.end()) {
            freq[i] = 1;
        } else {
            freq[i]++;
        }
    }

    auto cmpNode = [](Node* a, Node* b) {
        if (a->count != b->count) {
            return a->count > b->count;
        }

        if ((a->c == static_cast<char>(255)) != (b->c == static_cast<char>(255))) {
            return false;
        }

        return a->c < b->c;
    };

    std::priority_queue<std::shared_ptr<Node>, std::vector<std::shared_ptr<Node>>, decltype(cmpNode)> pq(cmpNode);

    for (auto i : freq) {
        pq.push(std::make_shared<Node>(Node{.c = i.first, .count = i.second, .left = nullptr, .right = nullptr}));
    }

    while (pq.size() > 1) {
        auto left = pq.top();
        pq.pop();

        auto right = pq.top();
        pq.pop();

        pq.push(std::make_shared<Node>(Node{ .c = static_cast<char>(255), .count = left->count + right->count, .left = left, .right = right }));
    }

    std::queue<std::shared_ptr<CodeNode>> q;

    q.push(std::make_shared<CodeNode>(CodeNode{ .node = pq.top(), .code = std::string() }));

    std::map<char, std::string> codes;

    while (!q.empty()) {
        auto e = q.front();
        q.pop();

        if (!e->node->left && !e->node->right) {
            codes[e->node->c] = e->code;
            continue;
        }

        if (e->node->left) {
            q.push(std::make_shared<CodeNode>(CodeNode{ .node = e->node->left, .code = e->code + "0" }));
        }

        if (e->node->right) {
            q.push(std::make_shared<CodeNode>(CodeNode{ .node = e->node->right, .code = e->code + "1" }));
        }
    }

    for (auto i : codes) {
        std::cout << "'" << i.first << "'" << ": " << i.second << std::endl;
    }

    return 0;
}
